### PR TITLE
TVMC: Add new text/relay frontend

### PR DESCRIPTION
**Pull request by PhilippvK**
This feature enables passing a textural representation of a relay module to the tvmc command line.

Example: `tvmc compile relay.txt --target c --runtime=crt --executor=aot --executor-aot-unpacked-api=1 --pass-config tir.disable_vectorize=1 -f mlf`

Currently it is not possible to supply parameters as it is mainly intended to be used for testing certain relay functions or operators. In the future (with minor changes to the tvmc frontend api) params could be passed via an additional i.e. `params.bin` file

This commit also adds minimal unit testing of the added feature.

### CODE REVIEW

**Comment by PhilippvK** [comment]
@masahi @gromero @leandron @jwfromm Please let me know if this is a meaningful addition to the TVMC command line interface. I am sure we could also agree on a way to provide actual values for the module parameters.

**Review by leandron — CHANGES_REQUESTED** [review]
In general the functionality looks good. I have reservations about calling it "text" frontend. My suggestion is to refer to it as "Relay" frontend, so that it is more specific. See my suggestions below, and if you agree, adjust in the other places affected, for example, log messages.

**Comment by PhilippvK** [comment]
@leandron Thank you for your feedback. I already considered using the name `relay` instead. I just though `RelayFrontend` might easy be confused with the implementations in `tvm.relay.frontend.*`.

What do you think about adding a way to provide custom (non-constant) data for the weights? This would involve small changes to the TVMC command line as more that one input file (e.g. `mod.relay` and `mod.params`) would need to be allowed in this special case.

**Comment by leandron** [comment]
> @leandron Thank you for your feedback. I already considered using the name `relay` instead. I just though `RelayFrontend` might easy be confused with the implementations in `tvm.relay.frontend.*`.
> 

I think the package makes it clear as a distinction, so I'd be in favour or renaming `TextFrontend` to `RelayFrontend`, given it is located at `tvm.driver.tvmc.frontend.RelayFrontend`, as a module.

> What do you think about adding a way to provide custom (non-constant) data for the weights? This would involve small changes to the TVMC command line as more that one input file (e.g. `mod.relay` and `mod.params`) would need to be allowed in this special case.

Not sure, it feels like we should think about a file format that incorporate all these for Relay, so that we bundle all that is needed in a single file to be used and reused in TVM, rather making it a special case in tvmc? cc @areusch

File: python/tvm/driver/tvmc/frontends.py
Comment (areusch) [inline python/tvm/driver/tvmc/frontends.py:313]: any thoughts on supplying data in e.g. `.npz`?

**Review by areusch — COMMENTED** [review]
thanks @PhilippvK, i agree with calling it `relay` frontend but am very supportive of adding this!

File: python/tvm/driver/tvmc/frontends.py
Comment (areusch) [reply python/tvm/driver/tvmc/frontends.py:313]: oh i just read your commit note more clearly. shall we instead add a warning perhaps so it's clear to the user how the interface is expected to work?

**Comment by PhilippvK** [comment]
Thanks for your feedback. I renamed all instances of `test` to `relay`.

The recommended approach for bundling the relay format with the parameters/weights wouldn't be solved in this PR, right?

As recommended by @areusch I could add a warning like `Parameters in the module will be initialized with ones.`. Any ideas for a different message?

**Comment by PhilippvK** [comment]
@leandron I totally agree. I will start a thread in the discussion forum in the next few days.

**Comment by leandron** [comment]
This failed in CI due to what is being fixed in https://github.com/apache/tvm/pull/11007. Once that gets merged, I think we can re-trigger CI here to get this merged.

**Review by leandron — APPROVED** [review]
LGTM, thanks!

**Comment by PhilippvK** [comment]
@leandron @areusch CI has successfully passed. Feel free to merge.

**Comment by PhilippvK** [comment]
I just ran into a few issues with the implementation when providing more complex models. I will have a look at but, but please do not merge this in its current state. Sorry for the inconvenience.

**Comment by areusch** [comment]
@PhilippvK we could also merge and you could submit some fixes. could you elaborate?

**Comment by PhilippvK** [comment]
@areusch 

These are the two issues I ran into:

1. As the frontend is not able to distinguish between real model inputs and parameters, I propagated all of the relay inputs with constant values. This way the model inputs are removed due to Constant Folding which makes the resulting model less complex. **Current Workaround:** Use `--input-shapes "x:[1,2,3]"` etc. (if available) to get the names of the actual  inputs and skip them when generating the parameter values.

2. I wanted to experiment with some models which make use of `meta[relay.Constant]` for internal constant vectors. These relay models can only be parsed if the Relay model was exported via `ir_mod.astext()` (which uses `show_meta_data=True` by default). The `relay.txt` format used in the MLF archive is using `str(ir_mod)` does neither include the `#[metadata]` section nor the `#[version ...]` header. I added a checker to detect if one this information is missing.

@areusch  Feel free to have another look at the changes.

**Comment by areusch** [comment]
hm. for (1), it seems like we could do something like:
- add a parameter which you have to pass for Relay models indicating the inputs
- accept params.npz and treat everything not in there as inputs
- a combination of both

for (2), that seems like an oversight on Model Library Format part. it seems like we should make tvmc validate the version numbers in relay models. Perhaps we should change Model Library Format?

i suppose it's also a bit of a burden on users authoring relay to figure out the presently-supported version, and they probably always just mean "the current version." so maybe allowing some implicitness there makes sense.

@leandron what are your thoughts?

**Comment by PhilippvK** [comment]
@areusch 

> add a parameter which you have to pass for Relay models indicating the inputs

I currently already use `--input-shapes` to get the input names, ignoring the provided shapes. Adding new flag for only the names would be somehow redundant in my opinion.

> accept params.npz and treat everything not in there as inputs

Would maybe be a good idea for the future, but I think this is out-of scope for this PR.

> for (2), that seems like an oversight on Model Library Format part.

The thing is that exporting the full relay model including metadata can result in very lang files which might not be desirable for the `relay.txt` in the MLF. However I would prefer some consistency here having at least the proper header in there. 

> it seems like we should make tvmc validate the version numbers in relay models.

I can look into this.

**Pull request merged**
